[BUGFIX lts] Fix tag cycles in query parameters

pzuraq opened the pull request:
Currently, we wrap query parameters with `@dependentKeyCompat` manually
to ensure they work with the legacy router setup, which watches QPs
using observers. When a QP value is labeled as `@tracked`, we don't need
to replace it with `@dependentKeyCompat`, but since it's just a plain
native getter (from the decorator) we do anyways. This results in a tag
cycle being created, which can result in a negative performance impact,
as every render pass will be invalidated by the cycle and require a
subsequent full revalidation.

This bug would have been caught by our ALLOW_CYCLES logic, but that
logic was faulty - it was allowing too many cycles, anything that was
accessed via `get`. It should only have allowed cycles from computeds
themselves.

This PR fixes the bug with ALLOW_CYCLES, and the subsequent bug with
`@tracked` by only wrapping with `@dependentKeyCompat` if the value is
not `@tracked`. It also adds an assertion to prevent users from using
`@dependentKeyCompat` with `@computed` or `@tracked`.

Verified that the test failed after fixing ALLOW_CYCLES, and passed
again after the fix.

rwjblue commented:
@pzuraq - Can you review the open issues RE: QP + cycles, and cross link the ones this fixes?

chriskrycho commented on packages/@ember/object/compat.ts:
```diff
@@ -2,15 +2,21 @@ import { Meta } from '@ember/-internals/meta';
   let { get: originalGet } = desc;
 
+  assert(
+    'You attempted to use @computed on a property that already has been decorated with either @computed or @tracked. @dependentKeyCompat is only necessary for native getters that are not decorated with @computed.',
```
@pzuraq do you mean "You attempted to use @dependentKeyCompat..." here?

rwjblue replied:
Good catch, fixed!

chancancode commented:
🙌 https://github.com/emberjs/ember.js/issues/18981

chancancode commented:
We should probably add a QP test for this?

rwjblue commented:
> We should probably add a QP test for this?

Yes, absolutely. That was why I was hoping we could cross link the related issues, and create test cases for them...

chancancode commented:
I think I have a small repro in twiddle, but maybe not small enough for a test case

pzuraq commented:
Note that there was already a test case, which was passing mistakenly due to the ALLOW_CYCLES logic being faulty. Fixing that logic caused the test to fail, and then the rest of the PR fixed it. Noted this in the original PR description.

@rwjblue did link me to a twiddle with an additional failure that somehow was caused by using the `{{log}}` helper, I added that as a test case as well.

Commit:
[BUGFIX lts] Fix tag cycles in query parameters

Currently, we wrap query parameters with @dependentKeyCompat manually
to ensure they work with the legacy router setup, which watches QPs
using observers. When a QP value is labeled as @tracked, we don't need
to replace it with @dependentKeyCompat, but since it's just a plain
native getter (from the decorator) we do anyways. This results in a tag
cycle being created, which can result in a negative performance impact,
as every render pass will be invalidated by the cycle and require a
subsequent full revalidation.

This bug would have been caught by our ALLOW_CYCLES logic, but that
logic was faulty - it was allowing too many cycles, anything that was
accessed via get. It should only have allowed cycles from computeds
themselves.

This PR fixes the bug with ALLOW_CYCLES, and the subsequent bug with
@tracked by only wrapping with @dependentKeyCompat if the value is
not @tracked. It also adds an assertion to prevent users from using
@dependentKeyCompat with @computed or @tracked.

Verified that the test failed after fixing ALLOW_CYCLES, and passed
again after the fix.

rwjblue reviewed (approved):
Rebased on top of all the prettier/eslint changes that just landed on master, and fixed up the error message issue that @chriskrycho pointed out.

Should be good to land once CI is happy...

Samsinite commented:
Does this fix the performance issue that was present when query params were used alongside tracked properties?
